### ISSUE TITLE
BUG: CategoricalDtype is not refresh after index categories set

### ISSUE THREAD

**Issue by Yikun** [pandas-dev/pandas#46820]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
pdf = pd.DataFrame(
          {
              "a": pd.Categorical([1, 2, 3, 1, 2, 3]),
          },
          index=pd.Categorical([10, 20, 30, 20, 30, 10], categories=[30, 10, 20], ordered=True),
      )
pidx = pdf.index
pidx.categories = ["z", "y", "x"]
pidx.dtype
```

### Issue Description

categories set failed

>>> pidx.dtype
CategoricalDtype(categories=[30, 10, 20], ordered=True)

### Expected Behavior

>>> pidx.dtype
CategoricalDtype(categories=['z', 'y', 'x'], ordered=True)

### Installed Versions

<details>

1.4.0+

</details>

**Comment by Yikun** [comment]
First bad commit: https://github.com/pandas-dev/pandas/commit/126a19d038b65493729e21ca969fbb58dab9a408#

also cc @jbrockmendel

**Comment by Yikun** [comment]
First bad commit: https://github.com/pandas-dev/pandas/commit/126a19d038b65493729e21ca969fbb58dab9a408#

also cc @jbrockmendel

**Comment by samukweku** [comment]
@Yikun it seems to work with `set_categories` right?

**Comment by Yikun** [comment]
@samukweku Yes, I can make it work as work ground:

```
pidx = pidx.set_categories(pidx.categories)
pdf.index = pidx
```

But I didn't need to do by this way before 1.4.0

**Comment by jreback** [comment]
you are trying to modify an immutable object - it's possible it accidentally worked before but in no way is this correct

**Comment by Yikun** [comment]
@jreback  Would we consider supporting `setters` as a way to simplify `set_categories`  in the future ?

**Comment by Yikun** [comment]
@jreback @jbrockmendel @samukweku Any thoughts?

**Comment by jbrockmendel** [comment]
we're actually looking at deprecating all the in-place category-setting behaviors xref #37643, so unless im misunderstanding what you're asking for, im -1 on adding this setter

**Comment by Yikun** [comment]
@jbrockmendel Thanks! And sorry for late reply.

Should we consider deprecating this [`setter`](https://github.com/pandas-dev/pandas/blob/main/pandas/core/arrays/categorical.py#L742-L743) or raise unsupported error? Because when the category is set (Yep, this is unexpected set to immutable object), the dtype is not refreshed, we are also no plan to fix it.

[1] https://github.com/pandas-dev/pandas/blob/main/pandas/core/arrays/categorical.py#L742-L743

**Comment by simonjayhawkins** [comment]
> you are trying to modify an immutable object

yes. it appears that the index is updated (which should probably raise) but the dtype attribute is not updated.

`pidx = pdf.index
print(pidx)
pidx.categories = ["z", "y", "x"]
print(pidx)
print(repr(pidx.dtype))
```
```
CategoricalIndex([10, 20, 30, 20, 30, 10], categories=[30, 10, 20], ordered=True, dtype='category')
CategoricalIndex(['y', 'x', 'z', 'x', 'z', 'y'], categories=['z', 'y', 'x'], ordered=True, dtype='category')
CategoricalDtype(categories=[30, 10, 20], ordered=True)
```

> First bad commit: [126a19d](https://github.com/pandas-dev/pandas/commit/126a19d038b65493729e21ca969fbb58dab9a408)

what code sample was used and what was the behavior change? I'm seeing the same behavior with the snippet above back to pandas 1.0.5

in pandas 0.25.3, it raised `AttributeError: can't set attribute`

**Comment by jbrockmendel** [comment]
categories.setter was removed in #49321, closing.
